### PR TITLE
Ensure that patterns are escaped

### PR DESCRIPTION
**Pull request by mcataford**
## Description

This quick change fixes a critical issue that arises when you try to filter using a pattern that contains regexp delimiters. Without it, trying to filter using a pattern that contains Regexp delimiters (i.e. `(`) crashes the shell.

### Steps to reproduce

- Open up the regular shell and invoke `hh`
- Search using a pattern that contains a parenthesis (or any other regexp-reserved character)
- See crash due to invalid regexp sequence
- Using this change, repeat.
- See no crash since the delimiters are escaped and considered to be the raw characters

### CODE REVIEW

**Comment by Akuli** [comment]
Regex searching is a feature, not a bug. For example, you're supposed to be able to type `^from` to find everything that starts with `from`, rather than everything that literally contains `^from`. This PR escapes that to the regex `r'\^from` which matches strings containing `^from`.

I think the correct fix would be to handle errors coming from bad regexes and display an error message in the UI.

**Comment by mcataford** [comment]
Interesting, it certainly came as an unexpected failure when trying to match entries with those characters; would an acceptable middle ground be to ignore bad regexps and treat them as the characters themselves? I.e. always consider the absolute string, but if it contains a valid regexp, then also consider that as an OR

That way, we can avoid the failure while still behaving as expected (i.e. if a valid regexp is typed, then use it, but otherwise use the filter query as-is).

**Comment by Akuli** [comment]
I think we should consider what each of these is supposed to do:
- `^from`: search for strings that start with `from`, strings that literally contain `^from` or both?
- `a(b)`: search for strings that contain `a(b)`, strings that contain `ab` or both?
- `a(b`: search for strings that contain `a(b` or display an error message?

@adder46 What do you think?

**Comment by Akuli** [comment]
tl;dr: This PR will likely be closed because it's a part of a bigger problem that adder46 is fixing.

I talked with adder46. Turns out that `pyhstr` has a regex mode and a non-regex mode. The default is non-regex mode, which is likely what you are using. However, `get_substring_indexes` is not aware of this difference. Also, `get_substring_indexes` is not aware of the case-sensitive and case-insensitive modes.

The correct solution is not to copy/paste the handling of different modes to `get_substring_indexes`, but instead modify the code containing the mode handling to also find the substring indexes, and then use the stored substring indexes in `get_substring_indexes`. I asked adder46 to implement this.

**Comment by mcataford** [comment]
That makes sense, feel free to close / I'll do it myself when I come back around. 👍

On September 19, 2020 12:54:03 p.m. EDT, Akuli <notifications@github.com> wrote:
>tl;dr: This PR will likely be closed because it's a part of a bigger
>problem that adder46 is fixing.
>
>I talked with adder46. Turns out that `pyhstr` has a regex mode and a
>non-regex mode. The default is non-regex mode, which is likely what you
>are using. However, `get_substring_indexes` is not aware of this
>difference. Also, `get_substring_indexes` is not aware of the
>case-sensitive and case-insensitive modes.
>
>The correct solution is not to copy/paste the handling of different
>modes to `get_substring_indexes`, but instead modify the code
>containing the mode handling to also find the substring indexes, and
>then use the stored substring indexes in `get_substring_indexes`. I
>asked adder46 to implement this.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/adder46/pyhstr/pull/2#issuecomment-695331415

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by adder46** [comment]
Sorry for not replying earlier. Anyways, Akuli spotted another bug. I'll see what I can do, if the patch comes in the meanwhile, I'll be glad to merge it.

**Comment by adder46** [comment]
Hm, I just tried to simply handle the error with the try/except block, and simply `pass` when `re.error` occurs. However, that didn't quite work out as expected. I'm hoping there has to be a simpler solution.

**Comment by Akuli** [comment]
If you don't like finding the substring indexes right away when searching the history, you could also convert the search string to a compiled regex object with `re.compile` in all of the different modes, like this:
- If in case insensitive mode, use `re.IGNORECASE` flag
- If not in regex mode, use `re.escape`

So now, any code that just uses the regex object is automatically aware of case sensitive/insensitive modes and regex/non-regex modes, because the code that creates the regex object handles all that. Then you would just use the regex object when searching and when finding substring indexes.

**Comment by Akuli** [comment]
adder's fixes are in #3
